### ISSUE TITLE
EpochsStream acquisition_delay for separate event stream

### ISSUE THREAD

**Issue by mzgsxs** [mne-tools/mne-lsl#369]
Hi I'm trying to understand and use EpochsStream class for my work. 

```python
epochs = EpochsStream(
    stream,
    bufsize=20,
    event_id=None,
    event_channels="LeftCue",
    event_stream=stream_events,
    tmin=-0.2,
    tmax=0.5,
    baseline=(None, 0),
    picks='eeg'
).connect(acquisition_delay=0.1)
epochs.info
```

It seems that if I use a separate event stream, connect with acquisition_delay < tmax, EpochsStream will always reject the events in ts_event. Is this behavior normal and expected or should this be considered as a bug?

https://github.com/mne-tools/mne-lsl/blob/1c60e82e70ed4f203a4a627b6b26ccc89864d9cc/src/mne_lsl/stream/epochs.py#L921

maybe it should be checked here, to see if it's smaller than tmax?

https://github.com/mne-tools/mne-lsl/blob/1c60e82e70ed4f203a4a627b6b26ccc89864d9cc/src/mne_lsl/stream/epochs.py#L317

thank you

**Comment by mzgsxs** [comment]
I think the problem is when outlet.push_sample() is called, it triggers EpochsStream's _acquire right away, this causes the drop of the event, as epoch data is not ready yet, however there is no additional _acquire planed ahead which causes this event to be dropped permenately

**Comment by mscheltienne** [comment]
That's too many inputs for the function signature:

```
def _prune_events(
    events: NDArray[np.int64],
    event_id: dict[str, int] | None,
    buffer_size: int,
    ts: NDArray[np.float64],
    last_ts: float | None,
    ts_events: NDArray[np.float64] | None,
    tmin_shift: float,
) -> NDArray[np.int64]:
```

Can you clarify a bit?

> I think the problem is when outlet.push_sample() is called, it triggers EpochsStream's _acquire right away, this causes the drop of the event, as epoch data is not ready yet, however there is no additional _acquire planed ahead which causes this event to be dropped permenately

If you provide `acquisition_delay=0.1`, it will call every 0.1s the method `_acquire` in the background thread (except if you are hogging CPU resources in the main thread, in which case the background thread will try to wake up and to call `_acquire` when the main thread gives him some headroom).

I'll have another look at your example tomorrow.

**Comment by mzgsxs** [comment]
thank you, sorry for my poor explanation 

> I think the problem is when outlet.push_sample() is called, it triggers EpochsStream's _acquire right away, this causes the drop of the event, as epoch data is not ready yet, however there is no additional _acquire planed ahead which causes this event to be dropped permenately

If I've understood your explanation correctly, acquisition_delay act as a timer and defines how often stream_events.n_new_samples is checked. if acquisition_delay is small,  _acquire is called almost right after stream_events.n_new_samples > 0, however at this point, if tmax is greater than the time passed from the tick/when _acquire was called (data from stream buffer does not have data for one complete epoch), the event is rejected and pruned. but since stream_events.n_new_samples has been reset to 0 after get_data call, if no new event occur after this and within stream's bufsize time (10s in the above example), further _acquire won't be able to get epoch data as it;s already out of the stream's buffer.

**Comment by mscheltienne** [comment]
But I think you are right that the description I just did is not what the code is doing.

- It's checking that there are new samples to work with, which makes sense, here: https://github.com/mne-tools/mne-lsl/blob/c274e56f496ab43353580acefa5cd73cd0f1276b/src/mne_lsl/stream/epochs.py#L434-L439
- It's retrieving the stream data and the event stream data with the public `get_data` method, here: https://github.com/mne-tools/mne-lsl/blob/c274e56f496ab43353580acefa5cd73cd0f1276b/src/mne_lsl/stream/epochs.py#L442
  and here: https://github.com/mne-tools/mne-lsl/blob/c274e56f496ab43353580acefa5cd73cd0f1276b/src/mne_lsl/stream/epochs.py#L464-L466

This public `get_data` call is resetting the number of new samples here: https://github.com/mne-tools/mne-lsl/blob/c274e56f496ab43353580acefa5cd73cd0f1276b/src/mne_lsl/stream/base.py#L615

I need to have a look tomorrow, but that might be the problem.

**Comment by mscheltienne** [comment]
Tests are passing here: https://github.com/mne-tools/mne-lsl/pull/371
Can you give a shot to the changes in this PR? You can install the wheel for your platform available at the end of the CI summary: https://github.com/mne-tools/mne-lsl/actions/runs/12075781960?pr=371
I'll update the PR tomorrow with a test reflecting your issue.

**Comment by mzgsxs** [comment]
thank you, I will try this first. 

> https://github.com/mne-tools/mne-lsl/actions/runs/12075781960?pr=371

**Comment by mzgsxs** [comment]
I think it now works as expected for my script, 
thank you very much for your help!

**Comment by mscheltienne** [comment]
Amazing, thanks for all the help in finding the bug!

**Comment by mscheltienne** [comment]
Released in 1.8
